File: python/src/euler022.py
import os


def solve():
    """
    Names scores
    Using names.txt, a 46K text file containing over five-thousand first names,
    begin by sorting it into alphabetical order. Then working out the alphabetical
    value for each name, multiply this value by its alphabetical position in the list
    to obtain a name score.
    For example, when the list is sorted into alphabetical order, COLIN, which is worth
    3 + 15 + 12 + 9 + 14 = 53, is the 522nd name in the list. So, COLIN would obtain
    a score of 522 × 53 = 27736.
    What is the total of all the name scores in the file?
    https://projecteuler.net/problem=22
    """
    data_path = os.path.join(os.path.dirname(__file__), '..', 'data', 'p022_names.txt')
    with open(data_path) as f:
        names = sorted(name.strip('"') for name in f.read().split(','))

    total = 0
    for i, name in enumerate(names, 1):
        score = sum(ord(c) - ord('A') + 1 for c in name)
        total += i * score
    return total

File: rust/src/euler022.rs
// https://projecteuler.net/problem=22
//
// Using a 46K text file containing over five-thousand first names, begin by sorting it
// into alphabetical order. Then working out the alphabetical value for each name,
// multiply this value by its alphabetical position in the list to obtain a name score.
//
// For example, when the list is sorted into alphabetical order, COLIN, which is worth
// 3 + 15 + 12 + 9 + 14 = 53, is the 938th name in the list. So, COLIN would obtain
// a score of 938 × 53 = 49714.
//
// What is the total of all the name scores in the file?
//
// Answer: 871198282

const NAMES_DATA: &str = include_str!("../../cpp/src/p022_names.txt");

pub fn names_scores() -> u64 {
    let mut names: Vec<&str> = NAMES_DATA
        .split(',')
        .map(|s| s.trim_matches('"'))
        .collect();
    names.sort_unstable();
    names
        .iter()
        .enumerate()
        .map(|(i, name)| {
            let alpha_val: u64 = name.bytes().map(|b| (b - b'A' + 1) as u64).sum();
            (i as u64 + 1) * alpha_val
        })
        .sum()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn euler_022() {
        assert_eq!(names_scores(), 871_198_282);
    }
}
